hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/storefiletracker/ModifyStoreFileTrackerProcedure.java

```diff
@@ -0,0 +1,271 @@
+    // there are 4 possible conditions:
+    // 1. The table or family has already made use of the dstSFT. In this way we just finish the
+    // procedure.
+    // 2. The table or family is not using the dstSFT but also not using migration SFT,
+    // then we just go to the MODIFY_STORE_FILE_TRACKER_MIGRATION state.
+    // 3. The table or family has already been using migration SFT and the dst SFT is what we
+    // expect, just go to MODIFY_STORE_FILE_TRACKER_FINISH.
+    // 4. The table or family is using migration SFT and the dst SFT is not what we
+    // expect, then need to schedule a MTP to change it to the dst SFT of the current migration
+    // SFT first, and then go to MODIFY_STORE_FILE_TRACKER_MIGRATION.
+    Class<? extends StoreFileTracker> clazz = StoreFileTrackerFactory.getTrackerClass(conf);
+    Class<? extends StoreFileTracker> dstSFTClass = StoreFileTrackerFactory.getTrackerClass(dstSFT);
+    if (clazz.equals(dstSFTClass)) {
+      return StoreFileTrackerState.ALREADY_FINISHED;
+    }
+    if (!MigrationStoreFileTracker.class.isAssignableFrom(clazz)) {
+      return StoreFileTrackerState.NEED_START_MIGRATION;
+    }
+    Class<? extends StoreFileTracker> currentDstSFT = StoreFileTrackerFactory
+      .getStoreFileTrackerClassForMigration(conf, MigrationStoreFileTracker.DST_IMPL);
+    if (currentDstSFT.equals(dstSFTClass)) {
+      return StoreFileTrackerState.NEED_FINISH_MIGRATION;
+    } else {
+      return StoreFileTrackerState.NEED_FINISH_PREVIOUS_MIGRATION_FIRST;
+    }
+  }
+
+  private final String getRestoreSFT(Configuration conf) {
+    Class<? extends StoreFileTracker> currentDstSFT = StoreFileTrackerFactory
+      .getStoreFileTrackerClassForMigration(conf, MigrationStoreFileTracker.DST_IMPL);
+    return StoreFileTrackerFactory.getStoreFileTrackerName(currentDstSFT);
+  }
+
+  protected abstract void preCheck(TableDescriptor current) throws IOException;
+
+  protected abstract Configuration createConf(Configuration conf, TableDescriptor current);
+
+  protected abstract TableDescriptor createRestoreTableDescriptor(TableDescriptor current,
+    String restoreSFT);
+
+  private Flow preCheckAndTryRestoreSFT(MasterProcedureEnv env) throws IOException {
+    // Checks whether the table exists
+    if (!env.getMasterServices().getTableDescriptors().exists(getTableName())) {
+      throw new TableNotFoundException(getTableName());
+    }
+    if (!isTableEnabled(env)) {
+      throw new TableNotEnabledException(tableName);
+    }
+    TableDescriptor current = env.getMasterServices().getTableDescriptors().get(tableName);
+    preCheck(current);
+    Configuration conf = createConf(env.getMasterConfiguration(), current);
+    StoreFileTrackerState state = checkState(conf, dstSFT);
+    switch (state) {
+      case NEED_FINISH_PREVIOUS_MIGRATION_FIRST:
+        TableDescriptor td = createRestoreTableDescriptor(current, getRestoreSFT(conf));
```

Comment:
I'm a bit confused here. Say we have three impls and the migration. Currently using impl1 and we submit migration to dst=impl2 (this triggers a MSFTP #1), followed by another migration to dst=impl3 (this triggers a MSFTP #2), then here we simply add a MTP to set impl2 tracker straight (no migration)? What if MSFTP #1 hasn't finished when this MTP starts run?

Reply:
What do you mean by "if MSFTP #1 hasn't finished when this MTP starts run"? The parent procedure can only resume execution when the child procedures are done. So there will be no overlap.

And typically, we only need to schedule two MTP, which is the flow starts from NEED_START_MIGRATION. First MTP changes the SFT implementation to MIGRATION, and the second MTP changes the SFT implementation to the dst implementation.

And why we need NEED_FINISH_PREVIOUS_MIGRATION_FIRST, is because we need to consider the situation where the SFT implementation is already MIGRATION, then we need to finish the MIGRATION by scheduling a MTP first. Then we can go back to the above logic to do the migration. Of course if the dst implementation the current MIGRATION is exactly what we want, then finish it is enough for us. That's why we have the logic in preCheckAndTryRestoreSFT.

I could try to add more comments here to explain the logic.

Reply:
Oh we already have comments in getState method. PTAL if it is clear enough for you @wchevreuil ? Thanks.

Reply:
> What do you mean by "if MSFTP #1 hasn't finished when this MTP starts run"? The parent procedure can only resume execution when the child procedures are done. So there will be no overlap.

What I meant is if two, subsequent, "change SFT" commands are triggered. Those would trigger separate "MSFTPs", with then independent child "MTPs". Or does a "MSFTP" hold a lock on the table, preventing this from happening?

Reply:
We hold exclusive lock on the table, same as MTP. We will release the lock between steps but there is no difference between MTP. I can not recall how this is handled or we just do not care, could dig more later. But anyway, there is no difference with the current MTP implementation.

Reply:
In TestModifyTableProcedure, there are several tests which run ModifyTableProcedure concurrently. I changed the log at the beginning of the executeFromState method of ModifyTableProcedure to debug, and find out that, although we scheduled the two MTP at almost the same time, the exeuction flow was always that, one MTP would wait for another to finish before executing...
Need to dig more. But anyway, I do not think this is a blocker issue for this PR?

Reply:
No, I think if no two MTPs can run concurrently, this is safe enough.